### ISSUE TITLE
lws minimal-http-server.c with curl http-post.c

### ISSUE THREAD

**Issue by smalls12** [warmcat/libwebsockets#1625]
I'm observing an issue with the minimal-http-server.c example from LWS and the http-post.c example from curl ( https://github.com/curl/curl/blob/master/docs/examples/http-post.c ).

Out of the box, the minimal-http-server.c will start and will receive a POST from http-post.c, albeit the server throws it away.

That's fine.

When I modify the http-post.c example to reuse the connection and send the POST for a second time, the server seems to spin out of control. The server appears to be reporting that it is receiving alot more messages than are being sent; maybe it can't see where the first POST ended and the second began?

The modification to http-post.c is to double the curl_easy_perform command below:

```
/* Perform the request, res will get the return code */
    res = curl_easy_perform(curl);
    /* Check for errors */
    if(res != CURLE_OK)
      fprintf(stderr, "curl_easy_perform() failed: %s\n",
              curl_easy_strerror(res));
```

Not sure if this is curl or lws, just looking to see if some obvious can be seen on this end.

Working on debug logs from libwebsockets to see if something else is reported.

Thanks for the help.

**Comment by smalls12** [comment]
Some output from libwebsockets.
Fyi, just on Ubuntu 16.04 x86

```
[2019/07/04 20:43:20:7510] INFO: Initial logging level 11
[2019/07/04 20:43:20:7510] INFO: Libwebsockets version: 3.1.0 v3.1.0
[2019/07/04 20:43:20:7510] INFO: IPV6 not compiled in
[2019/07/04 20:43:20:7510] INFO:  LWS_DEF_HEADER_LEN    : 4096
[2019/07/04 20:43:20:7510] INFO:  LWS_MAX_PROTOCOLS     : 5
[2019/07/04 20:43:20:7510] INFO:  LWS_MAX_SMP           : 1
[2019/07/04 20:43:20:7510] INFO:  sizeof (*info)        : 528
[2019/07/04 20:43:20:7511] INFO:  SYSTEM_RANDOM_FILEPATH: '/dev/urandom'
[2019/07/04 20:43:20:7511] INFO:  HTTP2 support         : available
[2019/07/04 20:43:20:7511] INFO: Using event loop: poll
[2019/07/04 20:43:20:7511] INFO: Default ALPN advertisment: h2,http/1.1
[2019/07/04 20:43:20:7511] INFO:  default timeout (secs): 20
[2019/07/04 20:43:20:7511] INFO:  Threads: 1 each 1024 fds
[2019/07/04 20:43:20:7511] INFO:  mem: context:          4912 B (816 ctx + (1 thr x 4096))
[2019/07/04 20:43:20:7511] INFO:  mem: http hdr rsvd:   5177344 B (1 thr x (4096 + 960) x 1024))
[2019/07/04 20:43:20:7511] INFO:  mem: pollfd map:       8192
[2019/07/04 20:43:20:7512] INFO:  mem: platform fd map:  8192 bytes
[2019/07/04 20:43:20:7512] INFO:  Compiled with OpenSSL support
[2019/07/04 20:43:20:7512] INFO:  SSL disabled: no LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT
[2019/07/04 20:43:20:7512] INFO:    mounting file://./mount-origin to /
[2019/07/04 20:43:20:7513] INFO: lws_vhost_bind_wsi: vh default: count_bound_wsi 1
[2019/07/04 20:43:20:7513] INFO:  Listening on port 7681
[2019/07/04 20:43:20:7513] INFO:  mem: per-conn:          544 bytes + protocol rx buf
[2019/07/04 20:43:20:7513] INFO: lws_cancel_service
[2019/07/04 20:43:20:7513] INFO: lws_protocol_init
```

Successful first POST
```
[2019/07/04 20:43:25:9003] INFO: lws_vhost_bind_wsi: vh default: count_bound_wsi 2
[2019/07/04 20:43:25:9003] INFO: lws_header_table_attach: wsi 0x15d5bd0: ah (nil) (tsi 0, count = 0) in
[2019/07/04 20:43:25:9003] INFO: _lws_create_ah: created ah 0x15d5e00 (size 4096): pool length 1
[2019/07/04 20:43:25:9003] INFO: lws_header_table_attach: did attach wsi 0x15d5bd0: ah 0x15d5e00: count 1 (on exit)
[2019/07/04 20:43:25:9003] INFO: rops_handle_POLLIN_listen: new wsi 0x15d5bd0: wsistate 0x20000015, role_ops h1
[2019/07/04 20:43:25:9004] INFO: lws_handshake_server: parsed count 123
[2019/07/04 20:43:25:9004] INFO: lws_select_vhost: vhost match to default based on port 7681
[2019/07/04 20:43:25:9004] INFO: lws_handshake_server: 0x15d5bd0: No upgrade
[2019/07/04 20:43:25:9004] INFO: Method: 'POST' (1), request for '/'
[2019/07/04 20:43:25:9004] INFO: no hit
[2019/07/04 20:43:25:9005] INFO: lws_issue_raw: ssl_capable_write (412) says 412
[2019/07/04 20:43:25:9005] INFO: lws_http_transaction_completed: wsi 0x15d5bd0
[2019/07/04 20:43:25:9005] INFO: __lws_header_table_detach: wsi 0x15d5bd0: ah 0x15d5e00 (tsi=0, count = 1)
[2019/07/04 20:43:25:9005] INFO: __lws_header_table_detach: nobody usable waiting
[2019/07/04 20:43:25:9005] INFO: _lws_destroy_ah: freed ah 0x15d5e00 : pool length 0
[2019/07/04 20:43:25:9005] INFO: __lws_header_table_detach: wsi 0x15d5bd0: ah 0x15d5e00 (tsi=0, count = 0)
[2019/07/04 20:43:25:9005] INFO: lws_http_action: 0x15d5bd0: LRS_BODY state set (0x20000016)
[2019/07/04 20:43:25:9006] INFO: lws_buflist_append_segment: len 24 first 1 (nil)
[2019/07/04 20:43:25:9006] INFO: lws_header_table_attach: wsi 0x15d5bd0: ah (nil) (tsi 0, count = 0) in
[2019/07/04 20:43:25:9006] INFO: _lws_create_ah: created ah 0x15d5e00 (size 4096): pool length 1
[2019/07/04 20:43:25:9006] INFO: lws_header_table_attach: did attach wsi 0x15d5bd0: ah 0x15d5e00: count 1 (on exit)
[2019/07/04 20:43:25:9006] INFO: HTTP_BODY_COMPLETION: 0x15d5bd0 (http-only)
[2019/07/04 20:43:25:9006] INFO: lws_http_transaction_completed: wsi 0x15d5bd0
[2019/07/04 20:43:25:9007] INFO: lws_buflist_aware_consume: draining rxflow: used 24, next 0
[2019/07/04 20:43:25:9007] INFO: lws_buflist_aware_consume: removed 0x15d5bd0 from dll_buflist
```

Second POST just spams this
```
[2019/07/04 20:43:30:9010] INFO: HTTP_BODY_COMPLETION: 0x15d5bd0 (http-only)
[2019/07/04 20:43:30:9010] INFO: lws_http_transaction_completed: wsi 0x15d5bd0
[2019/07/04 20:43:30:9010] INFO: lws_buflist_append_segment: len 147 first 1 (nil)
[2019/07/04 20:43:30:9010] INFO: HTTP_BODY_COMPLETION: 0x15d5bd0 (http-only)
[2019/07/04 20:43:30:9010] INFO: lws_http_transaction_completed: wsi 0x15d5bd0
[2019/07/04 20:43:30:9010] INFO: HTTP_BODY_COMPLETION: 0x15d5bd0 (http-only)
[2019/07/04 20:43:30:9010] INFO: lws_http_transaction_completed: wsi 0x15d5bd0
[2019/07/04 20:43:30:9011] INFO: HTTP_BODY_COMPLETION: 0x15d5bd0 (http-only)
[2019/07/04 20:43:30:9011] INFO: lws_http_transaction_completed: wsi 0x15d5bd0
[2019/07/04 20:43:30:9011] INFO: HTTP_BODY_COMPLETION: 0x15d5bd0 (http-only)
[2019/07/04 20:43:30:9011] INFO: lws_http_transaction_completed: wsi 0x15d5bd0
```

**Comment by lws-team** [comment]
It seems an lws problem dealing with the case the target url path didn't exist on the first POST.

It does the 404 processing before dealing with the body and loses track it should be getting a body.  You can reproduce it more clearly against the minimal-http-server-form-post example... sending pipelined POSTs to it as intended is OK

```
$ curl -v -L -d"text1=mytext1&send=mysend" http://127.0.0.1:7681/form1 http://127.0.0.1:7681/form1
```

Sending one POST to the wrong path is OK

```
$ curl -v -L -d"text1=mytext1&send=mysend" http://127.0.0.1:7681/x
```

But the combination gets into a state where the parser didn't consume anything before returning to the event loop, making it spin

```
$ curl -v -L -d"text1=mytext1&send=mysend" http://127.0.0.1:7681/x http://127.0.0.1:7681/x 
```

I pushed a patch that should fix it on v3.1-stable and master, please give it a try.

**Comment by smalls12** [comment]
Tested master, looks fixed.
Thanks!
